### PR TITLE
feat: Add tsconfig and transpile options

### PR DESCRIPTION
**Pull request by deekshas8**
Closes SAP/cloud-sdk-backlog#973.

- Add `tsConfig` option. Moved common code to `generator-common`.
- Remove `generateJs` option and replace with `transpile`. Also change default behavior
- Change default behavior of `packageJson`

- [x] I know which base branch I chose for this PR, as the default branch is `v2-main` now, which is not for v3 development.
- [x] If my change will be merged into the `main` branch (for v3), I've updated (V3-Upgrade-Guide.md)[./V3-Upgrade-Guide.md] in case my change has any implications for users updating to SDK v3

### CODE REVIEW

**Comment by marikaner** [comment]
Quick question: is it intended that "I know which base branch I chose for this PR, as the default branch is v2-main now, which is not for v3 development." is checked?

File: V3-Upgrade-Guide.md [hunk 1]
```diff
@@ -164,3 +165,12 @@ const myEntityApi = new MyEntityApi()
...
 To avoid this unintended usage of the constructor the visibility was changed to `private`.
 If you used the constructor directly please change your code to use the service function e.g. `myEntityService()` in the example above.
+
+### Update Transpilation options in Odata client generator
+
+The old `generateJs` option has been removed and instead replaced by `transpile`.
+By default, the OData generator will only generate TypeScript code.
+To generate JavaScript code, enable transpilation using the `--transpile` option.
+
+A new option, `tsConfig`, can be used to either pass a custom tsConfig configuration file or use a default config from the SDK.
```
Comment (marikaner) [inline V3-Upgrade-Guide.md:175]: [pp] The name of the file is `tsconfig.json`, maybe we should mention it explicitly.
```suggestion
A new option, `tsConfig`, can be used to either pass a custom `tsconfig.json` configuration file or use a default config from the SDK.
```

Comment (marikaner) [inline V3-Upgrade-Guide.md:175]: [pp] The name of the file is `tsconfig.json`, maybe we should mention it explicitly.
```suggestion
A new option, `tsConfig`, can be used to either pass a custom `tsconfig.json` configuration file or use a default config from the SDK.
```

Comment (marikaner) [inline V3-Upgrade-Guide.md:175]: [req] I assume this is a mistake from the ADR, but this should be called `tsconfig` not `tsConfig`, don't you think? As the name of the file this refers to is `tsconfig.json`...

File: knowledge-base/adr/0020-generator-options.md [hunk 2]
```diff
@@ -43,7 +43,7 @@ We should discuss the details of those defaults and make a conscious decision on
...
-| `--generatePackageJson`                                  | `--[no]-packageJson`             | Writes a default package.json, default true.                                                                | Writes a default package.json, default true (needs core dependency). Optionally in the future: Writes a custom package.json if passed. Keep boolean for now.                                                   |
+| `--generatePackageJson`                                  | `--[no]-packageJson`             | Writes a default package.json, default true.                                                                | Writes a default package.json, default false (needs core dependency). Optionally in the future: Writes a custom package.json if passed. Keep boolean for now.                                                   |
```
Comment (marikaner) [inline knowledge-base/adr/0020-generator-options.md:46]: [req] This says "core dependency" which is 1. quite cryptic (what is meant a dependency to the @sap-cloud-sdk/core package) and 2. totally outdated as there is no such package anymore. Could you correct this?

File: packages/generator/src/generator-options.ts [hunk 3]
```diff
@@ -84,9 +84,16 @@ export interface GeneratorOptions {
    */
   licenseInPackageJson?: string;
   /**
-   * Generates transpiled `.js`, `.js.map`, `.d.ts` and `.d.ts.map` files. When set to `false`, the generator will only generate `.ts` files.
+   * When enabled, generates transpiled `.js`, `.js.map`, `.d.ts` and `.d.ts.map` files.
```
Comment (marikaner) [inline packages/generator/src/generator-options.ts:87]: [req] The `.d.ts.map` files will be removed.

File: packages/generator/src/generator-options.ts [hunk 4]
```diff
@@ -257,9 +264,9 @@ export const generatorOptionsCli = {
   },
   packageJson: {
     describe:
-      'By default, the generator will generate a package.json file, specifying dependencies and scripts for compiling and generating documentation. When set to false, the generator will skip the generation of the package.json.',
+      'When enabled, a `package.json` that specifies dependencies and scripts for transpilation and documentation generation is generated.',
```
Comment (marikaner) [inline packages/generator/src/generator-options.ts:267]: [req] Documentation generation is not happening anymore, please remove.

File: packages/generator/src/generator-options.ts [hunk 5]
```diff
@@ -285,11 +292,17 @@ export const generatorOptionsCli = {
     deprecated:
       "Since v2.12.0. Use the 'include' option to add your own package.json file instead."
   },
-  generateJs: {
+  transpile: {
     describe:
-      'By default, the generator will also generate transpiled .js, .js.map, .d.ts and .d.ts.map files. When set to false, the generator will only generate .ts files.',
+      'Transpile the generated TypeScript code. When enabled a default `tsconfig.json` will be generated and used. It emits `.js`, `.js.map`, `.d.ts` and `.d.ts.map` files. To configure transpilation set `--tsconfig`.',
```
Comment (marikaner) [inline packages/generator/src/generator-options.ts:297]: [req] => as before: `.d.ts.map`

Comment (marikaner) [reply packages/generator/src/generator-options.ts:297]: [q] Here you wrote `--tsconfig` although currently it is called `tsConfig`. Does that work?

File: packages/openapi-generator/src/generator.ts [hunk 6]
```diff
@@ -40,8 +40,8 @@ import {
   getOriginalOptionsPerService,
   getRelPathWithPosixSeparator
 } from './options/options-per-service';
-import { tsconfigJson } from './options/tsconfig-json';
 import { sdkMetadata } from './sdk-metadata';
+import { tsconfigJson } from './options/tsconfig-json';
```
Comment (marikaner) [inline packages/openapi-generator/src/generator.ts:44]: [q] Can this be imported from './options`?

File: test-packages/test-services-odata-v2/generate-odata-services.ts [hunk 7]
```diff
@@ -7,7 +7,7 @@ const logger = createLogger('generate-odata-services');
 const generatorConfigOData = {
   forceOverwrite: true,
   overwrite: true,
-  generateJs: false,
+  transpile: false,
```
Comment (marikaner) [inline test-packages/test-services-odata-v2/generate-odata-services.ts:10]: [pp] You could shorten the list of properties by using the default.

**Review by marikaner — CHANGES_REQUESTED** [review]
Shouldn't this be "feat" + a changeset file and a compatibility note?

File: V3-Upgrade-Guide.md [hunk 8]
```diff
@@ -164,3 +165,12 @@ const myEntityApi = new MyEntityApi()
 the navigation properties are not correctly initialized leading to potential errors.
 To avoid this unintended usage of the constructor the visibility was changed to `private`.
 If you used the constructor directly please change your code to use the service function e.g. `myEntityService()` in the example above.
+
+### Update Transpilation options in Odata client generator
```
Comment (marikaner) [inline V3-Upgrade-Guide.md:169]: [req]
```suggestion
### Update Transpilation options in OData client generator
```

File: V3-Upgrade-Guide.md (hunk 1 above)
Comment (deekshas8) [reply V3-Upgrade-Guide.md:175]: In openapi generator it is called `tsConfig`, so I kept it like that. I'll adjust it in both places.

File: packages/openapi-generator/src/generator.ts (hunk 6 above)
Comment (deekshas8) [reply packages/openapi-generator/src/generator.ts:44]: yes

File: test-packages/test-services-odata-v2/generate-odata-services.ts (hunk 7 above)
Comment (deekshas8) [reply test-packages/test-services-odata-v2/generate-odata-services.ts:10]: I realized `generateJs` is set false here and then later to true in `generateOdata()`. Set `transpile` to true here (and only once)

**Comment by deekshas8** [comment]
> Quick question: is it intended that "I know which base branch I chose for this PR, as the default branch is v2-main now, which is not for v3 development." is checked?

Yes. My understanding of this statement is that 'I know what I am doing and have selected the correct base branch'.

File: .changeset/dirty-bats-unite.md [hunk 9]
```diff
@@ -0,0 +1,6 @@
+---
+'@sap-cloud-sdk/generator-common': major
+'@sap-cloud-sdk/generator': major
+---
+
+[New Functionality] Introduced options `transpile` and `tsconfig` to configure transpilation for a generated OData client
```
Comment (marikaner) [inline .changeset/dirty-bats-unite.md:6]: [pp] I'd use the more general plural here.
[req] End with a fullstop.

```suggestion
[New Functionality] Introduced options `transpile` and `tsconfig` to configure transpilation for generated OData clients.
```

File: V3-Upgrade-Guide.md [hunk 10]
```diff
@@ -164,3 +165,12 @@ const myEntityApi = new MyEntityApi()
...
 To avoid this unintended usage of the constructor the visibility was changed to `private`.
 If you used the constructor directly please change your code to use the service function e.g. `myEntityService()` in the example above.
+
+### Update Transpilation options in OData client generator
+
+By default, the OData generator will only generate TypeScript code.
+The `generateJs` option has been replaced with the `transpile` option.
+To generate JavaScript code, enable transpilation using the `transpile` option.
+
+A new option, `tsConfig`, can be used to either pass a custom `tsconfig.json` configuration file or use a default config from the SDK.
```
Comment (marikaner) [inline V3-Upgrade-Guide.md:175]: [req]
```suggestion
A new option, `tsconfig`, can be used to either pass a custom `tsconfig.json` configuration file or use a default config from the SDK.
```

**Review by marikaner — APPROVED** [review]
Looks good, I still left a few comments, though.

File: knowledge-base/adr/0020-generator-options.md [hunk 11]
```diff
@@ -43,7 +43,7 @@ We should discuss the details of those defaults and make a conscious decision on
...
 | `--serviceMapping`                                       | `--optionsPerService`            | Considers a per service configuration file. Shows a warning if none is given (differs from OData behavior). | Only generates a per service configuration, if a file path is provided. No warning if it is not provided. If provided, `packageName` and `directoryName` will be generated, but are not required to be set. \* |
 | `--tsConfig`                                             | `--tsConfig`                     | tsconfig.json file to overwrite the default "tsconfig.json".                                                | Writes a custom tsconfig.json if passed. Document that this should be used in combination with `transpile`.                                                                                                    |
```
Comment (deekshas8) [inline knowledge-base/adr/0020-generator-options.md:48]:
```suggestion
| `--tsconfig`                                             | `--tsconfig`                     | tsconfig.json file to overwrite the default "tsconfig.json".                                                | Writes a custom tsconfig.json if passed. Document that this should be used in combination with `transpile`.                                                                                                    |
```

**Pull request merged**